File: backend/Services/UpstreamApiService.cs
using System.Net.Http.Json;
using System.Text.Json;
using Microsoft.Extensions.Options;
using OperationNexus.Api.Configuration;
using OperationNexus.Api.Models.Upstream;

namespace OperationNexus.Api.Services;

public class UpstreamApiService : IUpstreamApiService
{
    private readonly HttpClient _httpClient;
    private readonly string _baseUrl;

    public UpstreamApiService(HttpClient httpClient, IOptions<UpstreamSettings> settings)
    {
        _httpClient = httpClient;
        _baseUrl = settings.Value.ApiUrl;
    }

    public async Task<(List<EmployeeDetail> Items, int TotalRecords)> GetEmployeesPagedAsync(string token, int skip, int take)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Post, $"{_baseUrl}employee/paged", token);
        request.Content = JsonContent.Create(new PagedRequest
        {
            Skip = skip,
            Take = take,
            Columns = BuildEmployeeColumns()
        });
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        var paged = await response.Content.ReadFromJsonAsync<PagedResponse>() ?? new();

        var items = paged.Payload.Select(row => new EmployeeDetail
        {
            UserId = GetInt(row, 0),
            FullName = GetString(row, 2),
            Email = GetString(row, 3),
            JobTitle = GetString(row, 4),
            MainSkillName = GetString(row, 5),
            OfficeName = GetString(row, 8),
        }).ToList();

        return (items, paged.FilteredRecordCount);
    }

    public async Task<EmployeeDetail> GetEmployeeDetailAsync(string token, int id)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Get, $"{_baseUrl}employee/get/{id}", token);
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<EmployeeDetail>() ?? new();
    }

    public async Task<List<EmployeeContract>> GetEmployeeContractsAsync(string token, int id)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Get, $"{_baseUrl}contract/{id}", token);
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<List<EmployeeContract>>() ?? new();
    }

    public async Task<List<EmployeeRate>> GetEmployeeRatesAsync(string token, int id)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Post, $"{_baseUrl}employee/{id}/rate", token);
        request.Content = JsonContent.Create(new PagedRequest
        {
            Skip = 0,
            Take = 100,
            Columns = BuildRateColumns()
        });
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        var paged = await response.Content.ReadFromJsonAsync<PagedResponse>() ?? new();

        return paged.Payload.Select(row => new EmployeeRate
        {
            AccountName = GetString(row, 0),
            ProjectName = GetString(row, 1),
            Rate = GetDecimal(row, 2),
            StartDate = GetString(row, 3),
        }).ToList();
    }

    public async Task<List<PersonaNote>> GetEmployeeNotesAsync(string token, int id)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Post, $"{_baseUrl}personanote/pagedByUser/{id}", token);
        request.Content = JsonContent.Create(new PagedRequest
        {
            Skip = 0,
            Take = 100,
            Columns = BuildNoteColumns()
        });
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        var paged = await response.Content.ReadFromJsonAsync<PagedResponse>() ?? new();

        return MapNoteRows(paged);
    }

    public async Task<(List<CandidateDetail> Items, int TotalRecords)> GetCandidatesPagedAsync(string token, int skip, int take)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Post, $"{_baseUrl}Candidate/paged", token);
        request.Content = JsonContent.Create(new PagedRequest
        {
            Skip = skip,
            Take = take,
            Columns = BuildCandidateColumns()
        });
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        var paged = await response.Content.ReadFromJsonAsync<PagedResponse>() ?? new();

        var items = paged.Payload.Select(row => new CandidateDetail
        {
            CandidateId = GetInt(row, 0),
            FullName = GetString(row, 1),
            Email = GetString(row, 11),
            MainSkill = GetString(row, 5),
            Seniority = GetString(row, 7),
            Country = GetString(row, 13),
        }).ToList();

        return (items, paged.FilteredRecordCount);
    }

    public async Task<CandidateDetail> GetCandidateDetailAsync(string token, int id)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Get, $"{_baseUrl}Candidate/{id}", token);
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadFromJsonAsync<CandidateDetail>() ?? new();
    }

    public async Task<List<PersonaNote>> GetCandidateNotesAsync(string token, int id)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Post, $"{_baseUrl}personanote/pagedByCandidate/{id}", token);
        request.Content = JsonContent.Create(new PagedRequest
        {
            Skip = 0,
            Take = 100,
            Columns = BuildNoteColumns()
        });
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        var paged = await response.Content.ReadFromJsonAsync<PagedResponse>() ?? new();

        return MapNoteRows(paged);
    }

    public async Task<byte[]> GetNoteFileAsync(string token, int noteId)
    {
        var request = CreateAuthorizedRequest(HttpMethod.Get, $"{_baseUrl}personanote/file/{noteId}", token);
        var response = await _httpClient.SendAsync(request);
        response.EnsureSuccessStatusCode();
        return await response.Content.ReadAsByteArrayAsync();
    }

    private HttpRequestMessage CreateAuthorizedRequest(HttpMethod method, string url, string token)
    {
        var request = new HttpRequestMessage(method, url);
        request.Headers.Add("x-sharepoint-token", token);
        return request;
    }

    private static List<ColumnDefinition> BuildEmployeeColumns()
    {
        return new List<ColumnDefinition>
        {
            new() { Name = "UserId", Label = "User ID", DataType = "numeric", IsKey = true, Filterable = false, Exportable = true, FilterOperator = "None", Visible = false },
            new() { Name = "Active", Label = "Active", FilterOperator = "Equals", FilterText = "Active", Filterable = false, Visible = false },
            new() { Name = "FullName", Label = "Name", Searchable = true, SortDirection = "Ascending", SortOrder = 1 },
            new() { Name = "Email", Label = "E-mail", Searchable = true },
            new() { Name = "JobTitle", Label = "Job Title", Searchable = true },
            new() { Name = "MainSkill", Label = "Main Skill", Searchable = true },
            new() { Name = "AdditionalSkills", Label = "Additional Skills", Searchable = true, Sortable = false },
            new() { Name = "FunctionalUnit", Label = "Functional Unit", Searchable = true },
            new() { Name = "Office", Label = "Office Location", Searchable = true },
            new() { Name = "ContractType", Label = "Contract Type", Searchable = true },
            new() { Name = "BusinessUnit", Label = "Business Unit", Searchable = true },
            new() { Name = "StartDate", Label = "Start Date", DataType = "datetimeutc", Searchable = false },
            new() { Name = "EnterpriseId", Label = "Enterprise ID", Searchable = true },
            new() { Name = "HrName", Label = "HR Partner", Searchable = true },
            new() { Name = "PeopleSuccessLead", Label = "People Success Lead", Searchable = true },
            new() { Name = "DateOfBirth", Label = "Date of Birth", DataType = "datetimeutc", Searchable = false, Sortable = false, Filterable = false },
            new() { Name = "AnniversaryDate", Label = "Anniversary Date", DataType = "datetimeutc", Searchable = false, Sortable = false, Filterable = false },
            new() { Name = "IsProjectBased", Label = "Is Project Based", Searchable = false },
        };
    }

    private static List<ColumnDefinition> BuildCandidateColumns()
    {
        return new List<ColumnDefinition>
        {
            new() { Name = "CandidateId", Label = "Actions", DataType = "numeric", IsKey = true, Filterable = false, Exportable = false, FilterOperator = "Equals" },
            new() { Name = "Candidate", Label = "Candidate", Searchable = true, SortDirection = "Ascending", SortOrder = 1 },
            new() { Name = "Recruiter", Label = "Recruiter", Searchable = true },
            new() { Name = "CandidateStatusName", Label = "Status", Filterable = false },
            new() { Name = "JobBoard", Label = "Job Board" },
            new() { Name = "Skills", Label = "Main Skill" },
            new() { Name = "AdditionalSkills", Label = "Additional Skill" },
            new() { Name = "Seniority", Label = "Seniority" },
            new() { Name = "SeniorityBand", Label = "Seniority Band" },
            new() { Name = "CoeCertifiedStatus", Label = "COE Certified Status" },
            new() { Name = "MobilePhone", Label = "Mobile Phone" },
            new() { Name = "Email", Label = "Email", Searchable = true },
            new() { Name = "SecondaryEmail", Label = "Secondary Email" },
            new() { Name = "Location", Label = "Location" },
            new() { Name = "StatusUpdate", Label = "Status Update", DataType = "datetimeutc" },
        };
    }

    private static List<ColumnDefinition> BuildRateColumns()
    {
        return new List<ColumnDefinition>
        {
            new() { Name = "AccountName", Label = "Account", Searchable = true },
            new() { Name = "ProjectName", Label = "Project", Searchable = true },
            new() { Name = "Rate", Label = "Rate", DataType = "numeric", Searchable = false },
            new() { Name = "StartDate", Label = "Start Date", DataType = "datetimeutc", Searchable = false },
        };
    }

    private static List<ColumnDefinition> BuildNoteColumns()
    {
        return new List<ColumnDefinition>
        {
            new() { Name = "PersonaNoteId", Label = "Note ID", DataType = "numeric", IsKey = true, Filterable = false, FilterOperator = "None", Visible = false },
            new() { Name = "NoteTypeName", Label = "Type", Searchable = true },
            new() { Name = "NoteContent", Label = "Content", Searchable = true },
            new() { Name = "FullName", Label = "Name", Searchable = true },
            new() { Name = "DateCreated", Label = "Created", DataType = "datetimeutc", Searchable = false },
            new() { Name = "Filename", Label = "File", Searchable = true },
        };
    }

    private static List<PersonaNote> MapNoteRows(PagedResponse paged)
    {
        return paged.Payload.Select(row => new PersonaNote
        {
            PersonaNoteId = GetInt(row, 0),
            NoteTypeName = GetString(row, 1),
            NoteContent = GetString(row, 2),
            FullName = GetString(row, 3),
            DateCreated = GetDateTime(row, 4),
            Filename = GetNullableString(row, 5),
        }).ToList();
    }

    private static string GetString(JsonElement[] row, int i) =>
        i < row.Length && row[i].ValueKind == JsonValueKind.String
            ? row[i].GetString() ?? string.Empty
            : i < row.Length ? row[i].ToString() : string.Empty;

    private static string? GetNullableString(JsonElement[] row, int i) =>
        i < row.Length && row[i].ValueKind == JsonValueKind.String
            ? row[i].GetString()
            : i < row.Length && row[i].ValueKind != JsonValueKind.Null ? row[i].ToString() : null;

    private static int GetInt(JsonElement[] row, int i) =>
        i < row.Length && row[i].ValueKind == JsonValueKind.Number
            ? row[i].GetInt32() : 0;

    private static decimal GetDecimal(JsonElement[] row, int i) =>
        i < row.Length && row[i].ValueKind == JsonValueKind.Number
            ? row[i].GetDecimal() : 0m;

    private static DateTime GetDateTime(JsonElement[] row, int i)
    {
        if (i < row.Length && row[i].ValueKind == JsonValueKind.String)
        {
            var str = row[i].GetString();
            if (DateTime.TryParse(str, out var dt))
                return dt;
        }
        return DateTime.MinValue;
    }
}

File: src/main/services/upstreamApiService.ts
import { getConfig } from '../config'
import { createLogger } from './logger'
import { fetchAuthorized, fetchPaged, type PagedResponse } from './upstream/upstreamApiClient'
import { getString, getNullableString, getInt, getDecimal, getBool, getDateTime, getNullableDateTime } from './upstream/upstreamRowParsers'
import { buildEmployeeColumns, buildCandidateColumns, buildRateColumns, buildNoteColumns, buildOpenPositionColumns, buildPresentedCandidateColumns, buildPrrColumns, buildPrrPresentationColumns, buildTeamCompositionColumns, buildBenchColumns } from './upstream/upstreamColumnDefs'
import { mapKeysToCamelCase } from './upstream/caseMapper'

export type { ColumnDefinition, PagedRequest, PagedResponse } from './upstream/upstreamApiClient'

const log = createLogger('UpstreamApi')

export interface EmployeeDetail {
  userId: number
  fullName: string
  email: string
  seniority: number
  mainSkillId: number
  countryId: number
  accountName: string
  jobTitle: string
  mainSkillName: string
  officeName: string
  functionalUnit: string
  businessUnit: string
}

export interface EmployeeContract {
  salary: number
  currencyCode: string
  startDate: string
  netSalary: number | null
  annualCost: number | null
}

export interface EmployeeRate {
  accountName: string
  projectName: string
  rate: number
  startDate: string
}

export interface CandidateDetail {
  candidateId: number
  firstName?: string
  lastName?: string
  fullName: string
  email?: string
  seniority?: number
  mainSkillId?: number
  countryId?: number
  coeCertifiedStatusId?: number
  statusUpdate?: string
  currentSalary?: number
  currentSalaryCurrency?: string
  desiredSalary?: string
  desiredSalaryCurrency?: string
  offer?: number
  mainSkill?: string
  country?: string
  seniorityText?: string
  coeCertifiedStatus?: string
  candidateStatusName?: string
  salaryCurrency?: string
}

export interface OpenPositionListItem {
  id: number
  account: string
  verticalIndustry: string
  coe: string
  practice: string
  stakeholder: string
  mainSkill: string
  countries: string
  seniorities: string
  availableRange: string
  status: string
  aging: number
  created: string
  readyDate: string
  lastModification: string
  sourcing: string
  replacement: boolean
  candidatesPresented: number
  lastDiscussionDate: string
  closedReason: string
  dateClosed: string | null
}

export interface OpenPositionDetail {
  recruitmentRequisitionId: number
  jobDescription: string
  companyName: string
  jobTitle: string
  mainSkillId: number
  mainSkillName: string
  seniorities: number[]
  countries: number[]
  seniority: string
  research: string
  active: string
  maximumRate: number | null
  minimumRate: number | null
  comments: string | null
  inOffice: boolean
  isReady: boolean
  isPromotion: boolean
  csu: string
  cs: string
  dateClosed: string | null
  additionalSkills: Array<{ tagId: number; tagName: string }>
  createdWithAssignmentsTool: boolean | null
}

export interface DiscussionCommentItem {
  commentId: number
  author: string
  date: string
  message: string
  parentCommentId: number | null
}

export interface PersonaNote {
  personaNoteId: number
  noteTypeName: string
  noteContent: string
  fullName: string
  dateCreated: string
  filename: string | null
}

export interface PresentedCandidateItem {
  candidateRequisitionId: number
  candidate: string
  candidateStatusName: string
  startDate: string
  skills: string
  rate: number | null
  isEmployee: boolean
  candidateId: number
}

export interface CandidateRequisitionDetailItem {
  candidateRequisitionId: number
  candidateFullName: string
  requisitionName: string
  requisitionStatus: number
  requisitionStatusId: number
  listFeedback: number[]
  comments: string
  rate: number
  actionDate: string
}

export interface PrrListItem {
  id: number
  employee: string
  account: string
  team: string
  mainSkill: string
  seniority: string
  transitionStatus: string
  transitionSubType: string
  location: string
  requestDate: string
  daysSinceLastInterview: string
  impact: string
  attritionRisk: string
  comments: string
}

export interface PrrPresentationItem {
  openPositionId: number
  account: string
  openPositionStatus: string
  location: string
  presentedOn: string
  candidateStatus: string
}

export interface TeamCompositionEntry {
  account: string
  team: string
  project: string
  role: string
  startDate: string
  endDate: string | null
}

function mapNoteRows(paged: PagedResponse): PersonaNote[] {
  return paged.payload.map(row => ({
    personaNoteId: getInt(row, 0),
    noteTypeName: getString(row, 1),
    noteContent: getString(row, 2),
    fullName: getString(row, 3),
    dateCreated: getDateTime(row, 4),
    filename: getNullableString(row, 5),
  }))
}

export const upstreamApiService = {
  async getEmployeesPaged(token: string, skip: number, take: number, signal?: AbortSignal): Promise<{ items: EmployeeDetail[]; totalRecords: number }> {
    const { upstream } = getConfig()
    const paged = await fetchPaged(`${upstream.apiUrl}employee/paged`, token, {
      skip, take, columns: buildEmployeeColumns(),
    }, signal)
    const items = paged.payload.map(row => ({
      userId: getInt(row, 0),
      fullName: getString(row, 2),
      email: getString(row, 3),
      jobTitle: getString(row, 4),
      mainSkillName: getString(row, 5),
      functionalUnit: getString(row, 7),
      officeName: getString(row, 8),
      businessUnit: getString(row, 10),
      seniority: 0,
      mainSkillId: 0,
      countryId: 0,
      accountName: '',
    }))
    log.debug('getEmployeesPaged', { skip, take, resultCount: items.length, totalRecords: paged.filteredRecordCount })
    return { items, totalRecords: paged.filteredRecordCount }
  },

  async getEmployeeDetail(token: string, id: number, signal?: AbortSignal): Promise<EmployeeDetail> {
    const { upstream } = getConfig()
    try {
      const raw = await fetchAuthorized<Record<string, unknown>>('GET', `${upstream.apiUrl}employee/get/${id}`, token, undefined, signal)
      return mapKeysToCamelCase<EmployeeDetail>(raw)
    } catch (err) {
      log.error(`getEmployeeDetail failed for id=${id}`, err instanceof Error ? err : new Error(String(err)), { upstreamId: id })
      throw err
    }
  },

  async getEmployeeContracts(token: string, id: number, signal?: AbortSignal): Promise<EmployeeContract[]> {
    const { upstream } = getConfig()
    try {
      const raw = await fetchAuthorized<Record<string, unknown>[]>('GET', `${upstream.apiUrl}contract/${id}`, token, undefined, signal)
      return Array.isArray(raw) ? raw.map(item => mapKeysToCamelCase<EmployeeContract>(item)) : []
    } catch (err) {
      log.error(`getEmployeeContracts failed for id=${id}`, err instanceof Error ? err : new Error(String(err)), { upstreamId: id })
      throw err
    }
  },

  async getEmployeeRates(token: string, id: number, signal?: AbortSignal): Promise<EmployeeRate[]> {
    const { upstream } = getConfig()
    try {
      const paged = await fetchPaged(`${upstream.apiUrl}employee/${id}/rate`, token, {
        skip: 0, take: 100, counter: 3, columns: buildRateColumns(),
      }, signal)
      return paged.payload.map(row => ({
        accountName: getString(row, 0),
        projectName: getString(row, 1),
        rate: getDecimal(row, 2),
        startDate: getString(row, 3),
      }))
    } catch (err) {
      log.error(`getEmployeeRates failed for id=${id}`, err instanceof Error ? err : new Error(String(err)), { upstreamId: id })
      throw err
    }
  },

  async getEmployeeNotes(token: string, id: number, signal?: AbortSignal): Promise<PersonaNote[]> {
    const { upstream } = getConfig()
    try {
      const paged = await fetchPaged(`${upstream.apiUrl}personanote/pagedByUser/${id}`, token, {
        skip: 0, take: 100, columns: buildNoteColumns(),
      }, signal)
      return mapNoteRows(paged)
    } catch (err) {
      log.error(`getEmployeeNotes failed for id=${id}`, err instanceof Error ? err : new Error(String(err)), { upstreamId: id })
      throw err
    }
  },

  async getEmployeeTeamComposition(token: string, id: number, signal?: AbortSignal): Promise<TeamCompositionEntry[]> {
    const { upstream } = getConfig()
    try {
      const paged = await fetchPaged(`${upstream.apiUrl}employee/${id}/composition`, token, {
        skip: 0, take: 100, columns: buildTeamCompositionColumns(),
      }, signal)
      return paged.payload.map(row => ({
        account: getString(row, 0),
        team: getString(row, 1),
        project: getString(row, 2),
        role: getString(row, 3),
        startDate: getString(row, 4),
        endDate: getNullableString(row, 5),
      }))
    } catch (err) {
      log.error(`getEmployeeTeamComposition failed for id=${id}`, err instanceof Error ? err : new Error(String(err)), { upstreamId: id })
      throw err
    }
  },

  async getCandidatesPaged(token: string, skip: number, take: number, year?: number, signal?: AbortSignal): Promise<{ items: CandidateDetail[]; totalRecords: number }> {
    const { upstream } = getConfig()
    const paged = await fetchPaged(`${upstream.apiUrl}Candidate/paged`, token, {
      skip, take, columns: buildCandidateColumns(year),
    }, signal)
    const items = paged.payload.map(row => ({
      candidateId: getInt(row, 0),
      fullName: getString(row, 1),
      candidateStatusName: getNullableString(row, 3) ?? undefined,
      email: getString(row, 11),
      mainSkill: getString(row, 5),
      seniorityText: getString(row, 7),
      country: getString(row, 13),
      coeCertifiedStatus: getNullableString(row, 9) ?? undefined,
      statusUpdate: getNullableDateTime(row, 14),
    }))
    log.debug('getCandidatesPaged', { skip, take, year, resultCount: items.length, totalRecords: paged.filteredRecordCount })
    return { items, totalRecords: paged.filteredRecordCount }
  },

  async getCandidateDetail(token: string, id: number, signal?: AbortSignal): Promise<CandidateDetail> {
    const { upstream } = getConfig()
    try {
      const raw = await fetchAuthorized<Record<string, unknown>>('GET', `${upstream.apiUrl}Candidate/${id}`, token, undefined, signal)
      return mapKeysToCamelCase<CandidateDetail>(raw)
    } catch (err) {
      log.error(`getCandidateDetail failed for id=${id}`, err instanceof Error ? err : new Error(String(err)), { upstreamId: id })
      throw err
    }
  },

  async getCandidateNotes(token: string, id: number, signal?: AbortSignal): Promise<PersonaNote[]> {
    const { upstream } = getConfig()
    try {
      const paged = await fetchPaged(`${upstream.apiUrl}personanote/pagedByCandidate/${id}`, token, {
        skip: 0, take: 100, columns: buildNoteColumns(),
      }, signal)
      return mapNoteRows(paged)
    } catch (err) {
      log.error(`getCandidateNotes failed for id=${id}`, err instanceof Error ? err : new Error(String(err)), { upstreamId: id })
      throw err
    }
  },

  async getNoteFile(token: string, noteId: number, signal?: AbortSignal): Promise<ArrayBuffer> {
    const { upstream } = getConfig()
    try {
      const response = await fetch(`${upstream.apiUrl}personanote/file/${noteId}`, {
        headers: { 'x-sharepoint-token': token },
        signal,
      })
      if (!response.ok) throw new Error(`Note file download failed: ${response.status}`)
      return response.arrayBuffer()
    } catch (err) {
      log.error(`getNoteFile failed for noteId=${noteId}`, err instanceof Error ? err : new Error(String(err)), { noteId })
      throw err
    }
  },

  async getOpenPositionsPaged(token: string, skip: number, take: number): Promise<{ items: OpenPositionListItem[]; totalRecords: number }> {
    const { upstream } = getConfig()
    const paged = await fetchPaged(`${upstream.apiUrl}op/paged/true/1/`, token, {
      skip, take, columns: buildOpenPositionColumns(),
    })
    const items = paged.payload.map(row => ({
      id: getInt(row, 1),
      account: getString(row, 2),
      verticalIndustry: getString(row, 3),
      coe: getString(row, 4),
      practice: getString(row, 5),
      stakeholder: getString(row, 6),
      mainSkill: getString(row, 7),
      status: getString(row, 8),
      countries: getString(row, 9),
      aging: row.length > 13 ? getInt(row, 13) : 0,
      seniorities: row.length > 14 ? getString(row, 14) : '',
      availableRange: row.length > 15 ? getString(row, 15) : '',
      created: row.length > 16 ? getString(row, 16) : '',
      readyDate: row.length > 17 ? getString(row, 17) : '',
      lastModification: row.length > 18 ? getString(row, 18) : '',
      sourcing: row.length > 19 ? getString(row, 19) : '',
      replacement: row.length > 20 && getBool(row, 20),
      candidatesPresented: row.length > 21 ? getInt(row, 21) : 0,
      lastDiscussionDate: row.length > 22 ? getString(row, 22) : '',
      closedReason: row.length > 23 ? getString(row, 23) : '',
      dateClosed: row.length > 21 ? getNullableString(row, 21) : null,
    }))
    log.debug('getOpenPositionsPaged', { skip, take, resultCount: items.length, totalRecords: paged.filteredRecordCount })
    return { items, totalRecords: paged.filteredRecordCount }
  },

  async getAllOpenPositionsPaged(token: string, skip: number, take: number, year?: number): Promise<{ items: OpenPositionListItem[]; totalRecords: number }> {
    const { upstream } = getConfig()
    const paged = await fetchPaged(`${upstream.apiUrl}op/paged/false//`, token, {
      skip, take, columns: buildOpenPositionColumns(year),
    })
    const items = paged.payload.map(row => ({
      id: getInt(row, 1),
      account: getString(row, 2),
      verticalIndustry: getString(row, 3),
      coe: getString(row, 4),
      practice: getString(row, 5),
      stakeholder: getString(row, 6),
      mainSkill: getString(row, 7),
      status: getString(row, 8),
      countries: getString(row, 9),
      aging: row.length > 13 ? getInt(row, 13) : 0,
      seniorities: row.length > 14 ? getString(row, 14) : '',
      availableRange: row.length > 15 ? getString(row, 15) : '',
      created: row.length > 16 ? getString(row, 16) : '',
      readyDate: row.length > 17 ? getString(row, 17) : '',
      lastModification: row.length > 18 ? getString(row, 18) : '',
      sourcing: row.length > 19 ? getString(row, 19) : '',
      replacement: row.length > 20 && getBool(row, 20),
      candidatesPresented: row.length > 21 ? getInt(row, 21) : 0,
      lastDiscussionDate: row.length > 22 ? getString(row, 22) : '',
      closedReason: row.length > 23 ? getString(row, 23) : '',
      dateClosed: row.length > 21 ? getNullableString(row, 21) : null,
    }))
    log.debug('getAllOpenPositionsPaged', { skip, take, year, resultCount: items.length, totalRecords: paged.filteredRecordCount })
    return { items, totalRecords: paged.filteredRecordCount }
  },

  async getOpenPositionDetail(token: string, id: number): Promise<OpenPositionDetail | null> {
    const { upstream } = getConfig()
    try {
      const raw = await fetchAuthorized<Record<string, unknown>>('GET', `${upstream.apiUrl}op/${id}`, token)
      return mapKeysToCamelCase<OpenPositionDetail>(raw)
    } catch (err) {
      log.error(`Failed to fetch open position detail for ${id}`, err instanceof Error ? err : new Error(String(err)))
      return null
    }
  },

  async getPresentedCandidates(token: string, positionId: number): Promise<PresentedCandidateItem[]> {
    const { upstream } = getConfig()
    try {
      const paged = await fetchPaged(`${upstream.apiUrl}op/pagedRequisition/${positionId}`, token, {
        skip: 0, take: 100, columns: buildPresentedCandidateColumns(),
      })
      return paged.payload.map(row => ({
        candidateRequisitionId: getInt(row, 0),
        candidate: getString(row, 1),
        candidateStatusName: getString(row, 2),
        startDate: getString(row, 3),
        skills: getString(row, 4),
        rate: row.length > 5 ? getDecimal(row, 5) : null,
        isEmployee: getBool(row, 6),
        candidateId: row.length > 8 ? getInt(row, 8) : 0,
      }))
    } catch (err) {
      log.error(`Failed to fetch presented candidates for position ${positionId}`, err instanceof Error ? err : new Error(String(err)))
      return []
    }
  },

  async getDiscussionComments(token: string, positionId: number): Promise<DiscussionCommentItem[]> {
    try {
      const url = `https://unocorpdiskus.azurewebsites.net/comments?topicCode=${positionId}&topicName=OPEN_POSITION`
      const response = await fetch(url, {
        headers: { 'Authorization': `Bearer ${token}` },
      })
      if (!response.ok) {
        log.warn(`Discussion API returned ${response.status} for position ${positionId}`)
        return []
      }
      const data = await response.json() as { comments?: Array<{ CommentId: number; Email: string; Comment: string; ParentCommentId: number | null; CreationDate: string }> }
      return (data.comments ?? []).map(c => ({
        commentId: c.CommentId,
        author: c.Email,
        date: c.CreationDate,
        message: c.Comment,
        parentCommentId: c.ParentCommentId,
      }))
    } catch (err) {
      log.error(`Failed to fetch discussion comments for position ${positionId}`, err instanceof Error ? err : new Error(String(err)))
      return []
    }
  },

  async getCandidateRequisitionDetail(token: string, candidateRequisitionId: number): Promise<CandidateRequisitionDetailItem | null> {
    const { upstream } = getConfig()
    try {
      const raw = await fetchAuthorized<Record<string, unknown>>(
        'GET',
        `${upstream.apiUrl}op/candidate/${candidateRequisitionId}`,
        token
      )
      return mapKeysToCamelCase<CandidateRequisitionDetailItem>(raw)
    } catch (err) {
      log.error(`Failed to fetch candidate requisition detail for ${candidateRequisitionId}`, err instanceof Error ? err : new Error(String(err)))
      return null
    }
  },

  async savePersonaNote(token: string, personId: number, noteType: string, fileName: string, fileContent: ArrayBuffer): Promise<number> {
    const { upstream } = getConfig()
    const formData = new FormData()
    formData.append('PersonId', personId.toString())
    formData.append('NoteTypeName', noteType)
    formData.append('File', new Blob([fileContent]), fileName)

    try {
      const response = await fetch(`${upstream.apiUrl}personanote/save`, {
        method: 'POST',
        headers: { 'x-sharepoint-token': token },
        body: formData,
      })
      if (!response.ok) throw new Error(`Save note failed: ${response.status}`)
      const raw = await response.json() as Record<string, unknown>
      const result = mapKeysToCamelCase<{ personaNoteId?: number }>(raw)
      const noteId = result.personaNoteId ?? 0
      log.info('savePersonaNote succeeded', { personId, noteType, fileName, noteId })
      return noteId
    } catch (err) {
      log.error(`savePersonaNote failed for personId=${personId}`, err instanceof Error ? err : new Error(String(err)), { personId, noteType, fileName })
      throw err
    }
  },

  async getPrrsPaged(token: string, skip: number, take: number): Promise<{ items: PrrListItem[]; totalRecords: number }> {
    const { upstream } = getConfig()
    const paged = await fetchPaged(
      `${upstream.apiUrl}ProjectTransition/paged?ProjectTransitionType=3&active=true`,
      token,
      { skip, take, columns: buildPrrColumns() }
    )
    const items = paged.payload.map(row => ({
      id: getInt(row, 0),
      employee: getString(row, 1),
      account: getString(row, 2),
      team: getString(row, 3),
      mainSkill: getString(row, 4),
      seniority: getString(row, 5),
      transitionStatus: getString(row, 6),
      transitionSubType: getString(row, 7),
      location: getString(row, 8),
      requestDate: getDateTime(row, 9),
      daysSinceLastInterview: getString(row, 10),
      impact: getString(row, 11),
      attritionRisk: getString(row, 12),
      comments: getString(row, 13),
    }))
    log.debug('getPrrsPaged', { skip, take, resultCount: items.length, totalRecords: paged.filteredRecordCount })
    return { items, totalRecords: paged.filteredRecordCount }
  },

  async getPrrPresentations(token: string, prrId: number): Promise<PrrPresentationItem[]> {
    const { upstream } = getConfig()
    try {
      const paged = await fetchPaged(
        `${upstream.apiUrl}projectTransition/presentations/${prrId}`,
        token,
        { skip: 0, take: 100, columns: buildPrrPresentationColumns(), counter: 1 }
      )
      return paged.payload.map(row => ({
        openPositionId: getInt(row, 0),
        account: getString(row, 1),
        openPositionStatus: getString(row, 2),
        location: getString(row, 3),
        presentedOn: getString(row, 4),
        candidateStatus: getString(row, 5),
      }))
    } catch (err) {
      log.error(`Failed to fetch PRR presentations for ${prrId}`, err instanceof Error ? err : new Error(String(err)))
      return []
    }
  },

  async getBenchEmployeeIds(token: string, signal?: AbortSignal): Promise<Set<number>> {
    const { upstream } = getConfig()
    const benchIds = new Set<number>()
    let skip = 0
    const take = 500

    try {
      while (true) {
        if (signal?.aborted) break
        const paged = await fetchPaged(
          `${upstream.apiUrl}Bench/paged/true`, token,
          { skip, take, columns: buildBenchColumns() }, signal
        )
        for (const row of paged.payload) {
          const userId = getInt(row, 0)
          if (userId > 0) benchIds.add(userId)
        }
        skip += paged.payload.length
        if (skip >= paged.filteredRecordCount || paged.payload.length === 0) break
      }
      log.info('Bench employee IDs loaded from upstream', { count: benchIds.size })
    } catch (err) {
      if (err instanceof Error && err.name === 'AbortError') throw err
      log.warn('Failed to load bench employee IDs from upstream — falling back to composition-only detection', {
        error: err instanceof Error ? err.message : String(err),
        partialCount: benchIds.size,
      })
    }
    return benchIds
  },
}
